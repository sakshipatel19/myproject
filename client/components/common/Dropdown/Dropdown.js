import React, { Fragment } from "react";
import Select, { components } from "react-select";
import "./Dropdown.scss";
import Icon from "../Icon";

const selectStyles = {
	control: (styles, { isDisabled }) => {
		return {
			borderStyle: "none",
			display: "flex",
			alignItems: "center",
			borderRadius: "7px",
			backgroundColor: isDisabled ? "#FAFAFA" : "#FFFFFF",
			height: "30px",
			paddingRight: "8px",
			boxShadow: isDisabled ? "none" :
				"0 1px 2px 0 rgba(0,67,132,0.04), 0 2px 4px 0 rgba(0,67,132,0.16)",
			border: isDisabled ? "1px solid #E7E7E7" : "1px solid transparent",
			"&:hover": {
				border: isDisabled ? "none" : "1px solid #0676ed"
			}
		};
	}
};

class Dropdown extends React.Component {
	onChange = selectedOption => {
		if (selectedOption.value !== this.props.selectedOption)
			this.props.onSelect(selectedOption.value);
	};

	createOptions = (options = []) =>
		options?.map(option => {
			return {
				value: option.value,
				label: this.createLabel(option)
			};
		});

	createLabel = (option = {}) => {
		const isSelected = this.props.selectedOption === option.value;
		let label;

		if (option.image)
			label = (
				<Fragment>
					<img src={option.image} className="select-option-image" />
					<span className="select-option-label-image">
						{option.label?.substring(0, 3)}
					</span>
				</Fragment>
			);
		else
			label = (
				<span className={`select-option-label`}>
					{option?.label?.length > 18 && isSelected
						? `${option?.label?.substring(0, 15)}...`
						: option?.label}
				</span>
			);

		return (
			<div
				className={`select-option-container ${isSelected ? "selected" : ""}`}
			>
				{label}
			</div>
		);
	};

	createLabelForSelectedValue = option => {
		let label;
		if (option?.image)
			label = (
				<div className={`select-option-container selected`}>
					<Fragment>
						<img src={option.image} className="select-option-image" />
						<span className="select-option-label-image">
							{option.label?.substring(0, 3)}
						</span>
					</Fragment>
				</div>
			);
		else
			label = (
				<div className={`select-option-container selected`}>
					<span className="select-option-label">
						{option?.label?.length > 21
							? `${option?.label?.substring(0, 18)}...`
							: option?.label}
					</span>
				</div>
			);
		return {
			value: option,
			label
		};
	};

	MenuComponent = props => {
		return (
			<components.Menu className={"dropdown-main-menu-container"} {...props}>
				{props.children}
				<Icon name={"upBlue"} size={10} iconClass={"dropdown-open-indicator"} />
			</components.Menu>
		);
	};

	DropdownIndicator = props => {
		return <Icon name={"down"} size={12} {...props} />;
	};

	OptionComponent = props => {
		return (
			<components.Option {...props} className={"dropdown-option"}>
				{props.children}
			</components.Option>
		);
	};

	MenuListComponent = props => {
		return (
			<components.MenuList
				{...props}
				className={"dropdown-menu-list"}
				data-test={"dropdown-menu-list"}
			>
				{props.children}
			</components.MenuList>
		);
	};

	filterOptions = () => {
		let newOptions = [...(this.props.options || [])];
		newOptions?.forEach((option, i) => {
			if (option.value === this.props.selectedOption) {
				newOptions.splice(i, 1);
				newOptions.unshift(option);
			}
		});
		return newOptions;
	};
	render() {
		const options = this.createOptions(this.filterOptions());
		let selectedOption = this.props.options?.find(
			e => e.value === this.props.selectedOption
		);
		selectedOption = this.createLabelForSelectedValue(selectedOption);

		return (
			<Select
				className={`dropdown-container`}
				data-test="dropdown-component"
				value={selectedOption}
				onChange={this.onChange}
				options={options}
				placeholder=""
				components={{
					IndicatorSeparator: null,
					DropdownIndicator:
						(options?.length > 1 && !this.props.isDisabled) ? this.DropdownIndicator : null,
					Menu: this.MenuComponent,
					Option: this.OptionComponent,
					MenuList: this.MenuListComponent
				}}
				styles={selectStyles}
				defaultValue={options[0]}
				isSearchable={false}
				isDisabled={this.props.isDisabled || options?.length <= 1}
			/>
		);
	}
}

export default Dropdown;
