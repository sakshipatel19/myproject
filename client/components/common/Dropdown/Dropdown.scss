.dropdown-container {
	width: 120px;
	cursor: pointer;
	color: #212129;

	.dropdown-main-menu-container {
		border: 1px solid #e7e7e7;
		border-radius: 7px;
		margin-top: -30px;
		background-color: #ffffff;
		box-shadow: 0 20px 60px -22px rgba(0, 67, 132, 0.3);
		overflow: hidden;
		.dropdown-open-indicator {
			top: 12px;
			right: 12px;
			box-sizing: initial;
			position: absolute;
		}
		.dropdown-menu-list {
			overflow-y: auto;
			overflow-x: hidden;
			padding: 0;

			&::-webkit-scrollbar {
				width: 5px;
				background-color: #fff;
			}
			&::-webkit-scrollbar-thumb {
				background-color: #ccc;
				border-radius: 10px;
			}
			.dropdown-option {
				background: #fff;
				padding: 7px 10px 8px 10px;
				cursor: pointer;
				&:hover {
					background-color: #def6ff;
				}
			}
			.select-option-label {
				padding-right: 14px;
			}
		}
	}

	.select-option-container {
		display: flex;
		align-items: center;
		cursor: pointer;
		.select-option-image {
			height: 14px;
			width: 24px;
		}
		.select-option-label,
		.select-option-label-image {
			font-size: 12px;
			font-family: "NotoSans-Regular";
			color: #212129;
		}
		.select-option-label-image {
			margin-left: 15px;
		}

		&.selected {
			.select-option-label,
			.select-option-label-image {
				font-family: "NotoSans-SemiBold";
			}
		}
	}
}

.dropdown-container-disabled {
	color: #858585;
}
