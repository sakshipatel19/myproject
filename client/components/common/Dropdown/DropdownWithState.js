import React, { Component } from "react";
import Dropdown from '.';

class DropdownWithState extends Component {

    state = {
        options: []
    }

    static getDerivedStateFromProps(nextProps, prevState) {
        if (nextProps.options !== prevState.options && nextProps?.options?.length > 0) {
            return { options: nextProps.options };
        }
        return null;
    }

    render() {
        return (
            <Dropdown
                options={this.state.options}
                onSelect={this.props.onSelect}
                selectedOption={this.props.selectedOption}
            />);
    }
}
export default DropdownWithState;