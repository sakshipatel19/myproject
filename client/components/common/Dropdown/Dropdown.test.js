import React from "react";
import { shallow } from "enzyme";

import Dropdown from ".";

const properties = {
	selectedOption: null,
	options: [
		{ value: 1, label: "test1" },
		{ value: 2, label: "test2" },
		{ value: 3, label: "test3" }
	],
	onSelect: () => {}
};

const setup = (props = {}, state = null) => {
	const wrapper = shallow(<Dropdown {...props} />);
	if (state) wrapper.setState(state);
	return wrapper;
};

const findByTestAttr = (wrapper, val) => {
	return wrapper.find(`[data-test="${val}"]`);
};

describe("Dropdown Load testing", () => {
	it("Dropdown Render correctly", () => {
		const wrapper = setup(properties);

		expect(wrapper).toMatchSnapshot();
	});

	it("dropdown renders without error", () => {
		const wrapper = setup({ ...properties });
		const dropdownComponent = findByTestAttr(wrapper, "dropdown-component");
		expect(dropdownComponent.length).toBe(1);
	});

	// it("click on dropdown to open menu list", () => {
	// 	const wrapper = setup({ ...properties });
	// 	const dropdownComponent = findByTestAttr(wrapper, "dropdown-component");
	// 	dropdownComponent.simulate("click");
	// 	const menuComponent = findByTestAttr(wrapper, "dropdown-menu-list");
	// 	expect(menuComponent.length).toBe(1);
	// });
});
