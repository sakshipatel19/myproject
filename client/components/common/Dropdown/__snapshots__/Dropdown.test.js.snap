// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Load testing Dropdown Render correctly 1`] = `
<StateManager
  className="dropdown-container"
  components={
    Object {
      "DropdownIndicator": [Function],
      "IndicatorSeparator": null,
      "Menu": [Function],
      "MenuList": [Function],
      "Option": [Function],
    }
  }
  data-test="dropdown-component"
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={
    Object {
      "label": <div
        className="select-option-container "
      >
        <span
          className="select-option-label"
        >
          test1
        </span>
      </div>,
      "value": 1,
    }
  }
  isDisabled={false}
  isSearchable={false}
  onChange={[Function]}
  options={
    Array [
      Object {
        "label": <div
          className="select-option-container "
        >
          <span
            className="select-option-label"
          >
            test1
          </span>
        </div>,
        "value": 1,
      },
      Object {
        "label": <div
          className="select-option-container "
        >
          <span
            className="select-option-label"
          >
            test2
          </span>
        </div>,
        "value": 2,
      },
      Object {
        "label": <div
          className="select-option-container "
        >
          <span
            className="select-option-label"
          >
            test3
          </span>
        </div>,
        "value": 3,
      },
    ]
  }
  placeholder=""
  styles={
    Object {
      "control": [Function],
    }
  }
/>
`;
